FLUME-3085: HDFS Sink can skip flushing some BucketWriters, might lead to data loss

adenes opened the pull request:
This commit fixes the issue when in `HDFSEventSink.process()` a `BucketWriter.append()` call threw a `BucketClosedException` then the newly created `BucketWriter` wasn't flushed after the processing loop.

adenes committed:
FLUME-3085: HDFS Sink can skip flushing some BucketWriters, might lead to data loss

This commit fixes the issue when in HDFSEventSink.process() a BucketWriter.append()
call threw a BucketClosedException then the newly created BucketWriter wasn't
flushed after the processing loop.

adenes commented:
Some comments for the implementation:
- I replaced the `writers` list to a set. Actually to a `LinkedHashSet` in order to be consistent with the `List` implementation regarding to the ordering of the elements. (Although the logic doesn't depend on the ordering.)
- The `writers` variable was moved into the `try-catch` block as it isn't used outside of it.
- I moved the `writers.add(bucketWriter)` call to right after the `append()` call. It means that the first `BucketWriter` won't be added if it has been closed already (so a new `BucketWriter` was created), but it isn't necessary as flushing a closed `BucketWriter` does nothing (and flush happened during the close, anyway).
- I had to change the visibility of the `initializeBucketWriter()` method to package private to be able to mock the `BucketWriter`s in the test.

Tests in `flume-ng-sinks/flume-hdfs-sink` pass, checkstyle as well.

simonati commented on flume-ng-sinks/flume-hdfs-sink/src/test/java/org/apache/flume/sink/hdfs/TestHDFSEventSink.java:
```diff
@@ -1545,4 +1549,61 @@ private void testRetryRename(boolean closeSucceed)
+    // After processing the events the only BucketWriter in the sfWriters map is
+    // the one which was created after the first one threw the BucketClosedException.
+    // It is expected that its flush() method was called exactly once.
+    Map<String, BucketWriter> bucketWriterMap = sink.getSfWriters();
```
Instead of accessing sfwriters you can collect the created BucketWriters from your test extension of HDFSEventSink. This way you could check that whatever bucketwriter was created is actually flushed/closed. sfwriters list might not have reference to every created bucketwriter (lookup path is common so the newly created bw kicks out the closed one on line 406).

simonati commented on flume-ng-sinks/flume-hdfs-sink/src/main/java/org/apache/flume/sink/hdfs/HDFSEventSink.java:
```diff
@@ -415,6 +410,9 @@ public void run(String bucketPath) {
         }
+
+        // track the buckets getting written in this transaction
+        writers.add(bucketWriter);
```
Looks good to also capture the bucketwriter created on line 406. Since the original (line 389 or 393) one threw a closed exception that means that we don't have to flush the original one only the newly created one (close implicitly contains flush as well)

simonati commented on flume-ng-sinks/flume-hdfs-sink/src/test/java/org/apache/flume/sink/hdfs/TestHDFSEventSink.java:
```diff
@@ -1545,4 +1549,61 @@ private void testRetryRename(boolean closeSucceed)
+    sink.setChannel(channel);
+    sink.start();
+
+    Transaction txn = channel.getTransaction();
```
nit: I would have stubbed the channel.take() instead of juggling around with this transaction.

simonati reviewed (changes requested):
Business logic looks good to me. 
Improvement requested on testing part only as it test a slightly different thing (sfWriter list vs all bucketwriters).
mvn test -pl flume-ng-sinks/flume-hdfs-sink passes.

adenes committed:
FLUME-3085: HDFS Sink can skip flushing some BucketWriters, might lead to data loss

Change based on the reviews: collect the mocked BucketWriters instead
of using the HDFSEventSink.sfWriters map.

adenes committed:
FLUME-3085: HDFS Sink can skip flushing some BucketWriters, might lead to data loss

Mock the channel to make the juggling with transaction uneccesarry.

adenes commented:
Thank you for the review @simonati, I have updated the change based on your comments, please let me know what you think.

simonati commented:
Hi @bessbd, @mpercy,

Could you please retrigger Travis CI? (build failed because of https://repo.maven.apache.org/maven2 had been unreachable)
I doesn't seem to have the necessary entitlement to do it. 

Cheers,
Attila

mpercy commented:
@simonati I don't know how to do that; try just re-pushing the patch.

simonati commented:
Hi @mpercy,
For future reference a workaround exists: amending the commit message and force pushing the last commit retriggered Travis CI. (And any meaningful new commit would do the same)

adenes commented:
@mpercy, @bessbd, could you please review this change and commit if you are ok with it? Thanks in advance.

simonati commented on flume-ng-sinks/flume-hdfs-sink/src/test/java/org/apache/flume/sink/hdfs/TestHDFSEventSink.java:
```diff
@@ -1545,4 +1553,62 @@ private void testRetryRename(boolean closeSucceed)
+    Context context = new Context(ImmutableMap.of("hdfs.path", testPath));
+    Configurables.configure(sink, context);
+
+    Channel channel = Mockito.spy(new MemoryChannel());
```
Ahh I see. MockChannel is in flume-ng-core test folder thus cannot be used as is. And a simple Mockito.mock(MemoryChannel.class) is not enough as beside take(), also getTransaction has to be implemented. It might need a later refactoring to move this Mock implementations to a shared module but for now I'm fine with this.

simonati reviewed (approved):
LGTM

mpercy commented on flume-ng-sinks/flume-hdfs-sink/src/main/java/org/apache/flume/sink/hdfs/HDFSEventSink.java:
```diff
@@ -415,6 +410,9 @@ public void run(String bucketPath) {
         }
+
+        // track the buckets getting written in this transaction
+        writers.add(bucketWriter);
```
nit: Can we keep the if-statement? I suppose add() on the same element multiple times would just replace the previous element in the set and end up as a no-op, but it seems clearer to me to keep the "contains" there, especially now that you've switched to a hash-based container and it would be an O(1) operation, e.g.

```
if (!writers.contains(bucketWriter)) {		
  writers.add(bucketWriter);		
}
```

adenes replied:
sure, will update the code

mpercy commented on flume-ng-sinks/flume-hdfs-sink/src/main/java/org/apache/flume/sink/hdfs/HDFSEventSink.java:
```diff
@@ -354,9 +354,9 @@ static CompressionCodec getCodec(String codecName) {
-    List<BucketWriter> writers = Lists.newArrayList();
     transaction.begin();
     try {
+      Set<BucketWriter> writers = new LinkedHashSet<>();
```
nit: I don't think we need ordering, how about just HashSet<>()?

adenes replied:
I'm not sure but I vaguely remember some issue regarding to the ordering (cc. @simonati) so I would keep the `LinkedHashSet`. On the other hand not tests break with `HashSet`.

mpercy replied:
OK

mpercy commented on flume-ng-sinks/flume-hdfs-sink/src/test/java/org/apache/flume/sink/hdfs/TestHDFSEventSink.java:
```diff
@@ -1545,4 +1553,62 @@ private void testRetryRename(boolean closeSucceed)
+            hdfsWriter, closeCallback));
+        try {
+          // create mock BucketWriters where the first append() succeeds but the
+          // the second call throws a BucketClosedException
```
I don't see any logic here to make this throw only on the second call. It seems like it would throw on every call. Maybe I'm missing something.

adenes replied:
Chaining the `do*` calls in Mockito's Stubber class ensures that the first call will call the real method and the second call throws the exception. 
For more details see the Stubber's javadoc: https://static.javadoc.io/org.mockito/mockito-core/1.9.0/org/mockito/stubbing/Stubber.html

mpercy commented on flume-ng-sinks/flume-hdfs-sink/src/test/java/org/apache/flume/sink/hdfs/TestHDFSEventSink.java:
```diff
@@ -1545,4 +1553,62 @@ private void testRetryRename(boolean closeSucceed)
+    sink.process();
+
+    // It is expected that flush() method was called exactly once for every BucketWriter
+    for (BucketWriter bw : bucketWriters) {
```
I think it would be good to add some more assertions: how many events were taken from the channel, how many events were successfully written to HDFS, and how many bucketWriters there are.

adenes replied:
ok, will do.

mpercy reviewed (changes requested):
Great find. I just have a few mostly minor comments.

adenes commented:
thanks @mpercy for the review, I'll update the PR soon.

adenes committed:
FLUME-3085: HDFS Sink can skip flushing some BucketWriters, might lead to data loss

Update based on the code reviews:
- add if (!contains..) before writers.add() in HDFSEventSink
- add extra assertions to the test case

adenes commented:
Hi @mpercy, I have updated the pull request based on your review. Could you please have a look on it again? Thanks.

mpercy reviewed (approved):
+1 LGTM
